File: src/main/java/wraith/fabricaeexnihilo/compatibility/modules/MetaModule.java
package wraith.fabricaeexnihilo.compatibility.modules;

import org.jetbrains.annotations.NotNull;
import wraith.fabricaeexnihilo.api.compatibility.FabricaeExNihiloModule;
import wraith.fabricaeexnihilo.api.registry.*;

import java.util.ArrayList;
import java.util.List;

public class MetaModule implements FabricaeExNihiloModule {

    public static MetaModule INSTANCE = new MetaModule();

    private MetaModule() {}

    public List<FabricaeExNihiloModule> modules = new ArrayList<>();

    @Override
    public void registerCrucibleHeat(@NotNull CrucibleHeatRecipeRegistry registry) {
        modules.forEach(module -> module.registerCrucibleHeat(registry));
    }

    @Override
    public void registerCrucibleStone(@NotNull CrucibleRecipeRegistry registry) {
        modules.forEach(module -> module.registerCrucibleStone(registry));
    }

    @Override
    public void registerCrucibleWood(@NotNull CrucibleRecipeRegistry registry) {
        modules.forEach(module -> module.registerCrucibleWood(registry));
    }

    @Override
    public void registerOres(@NotNull OreRecipeRegistry registry) {
        modules.forEach(module -> module.registerOres(registry));
    }

    @Override
    public void registerSieve(@NotNull SieveRecipeRegistry registry) {
        modules.forEach(module -> module.registerSieve(registry));
    }

    @Override
    public void registerMesh(@NotNull MeshRecipeRegistry registry) {
        modules.forEach(module -> module.registerMesh(registry));
    }

    @Override
    public void registerCrook(@NotNull ToolRecipeRegistry registry) {
        modules.forEach(module -> module.registerCrook(registry));
    }

    @Override
    public void registerHammer(@NotNull ToolRecipeRegistry registry) {
        modules.forEach(module -> module.registerHammer(registry));
    }

    @Override
    public void registerWitchWaterWorld(@NotNull WitchWaterWorldRecipeRegistry registry) {
        modules.forEach(module -> module.registerWitchWaterWorld(registry));
    }

    @Override
    public void registerWitchWaterEntity(@NotNull WitchWaterEntityRecipeRegistry registry) {
        modules.forEach(module -> module.registerWitchWaterEntity(registry));
    }

}

File: src/main/kotlin/exnihilofabrico/compatibility/modules/MetaModule.kt
package exnihilofabrico.compatibility.modules

import exnihilofabrico.api.compatibility.IExNihiloFabricoModule
import exnihilofabrico.api.registry.*

object MetaModule: IExNihiloFabricoModule {

    val modules: MutableList<IExNihiloFabricoModule> = mutableListOf()

    init {
        modules.add(ExNihiloFabrico)
        // Check if an mod with a specific module is installed, then add its module to the list
    }

    override fun registerAlchemy(registry: IAlchemyRegistry) = modules.forEach { it.registerAlchemy(registry) }
    override fun registerCompost(registry: ICompostRegistry) = modules.forEach { it.registerCompost(registry) }
    override fun registerLeaking(registry: ILeakingRegistry) = modules.forEach { it.registerLeaking(registry) }
    override fun registerMilking(registry: IMilkingRegistry) = modules.forEach { it.registerMilking(registry) }
    override fun registerCrucibleHeat(registry: ICrucibleHeatRegistry) = modules.forEach { it.registerCrucibleHeat(registry) }
    override fun registerCrucibleStone(registry: ICrucibleRegistry) = modules.forEach { it.registerCrucibleStone(registry) }
    override fun registerCrucibleWood(registry: ICrucibleRegistry) = modules.forEach { it.registerCrucibleWood(registry) }
    override fun registerOres(registry: IOreRegistry) = modules.forEach { it.registerOres(registry) }
    override fun registerSieve(registry: ISieveRegistry) = modules.forEach { it.registerSieve(registry) }
    override fun registerMesh(registry: IMeshRegistry) = modules.forEach { it.registerMesh(registry) }
    override fun registerCrook(registry: IToolRegistry) = modules.forEach { it.registerCrook(registry) }
    override fun registerHammer(registry: IToolRegistry) = modules.forEach { it.registerHammer(registry) }
    override fun registerWitchWaterWorld(registry: IWitchWaterWorldRegistry) = modules.forEach { it.registerWitchWaterWorld(registry) }
    override fun registerWitchWaterEntity(registry: IWitchWaterEntityRegistry) = modules.forEach { it.registerWitchWaterEntity(registry) }
}